### PR TITLE
journal: add object sanity check to journal_file_move_to_object() (#6447)

### LINKED ISSUE #6447: journalctl --list-boots fails with v232-v234 again (was working in v231)
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd v233/v234

### Used distribution

company specific.

### In case of bug report: Expected behaviour you didn't see

journalctrl --list-boots shuold print all boot ids.

### In case of bug report: Unexpected behaviour you saw

journalctl --list-boots returns error:
Failed to determine boots: No data available

### In case of bug report: Steps to reproduce the problem

causing corrupted journal files with "reboot -f" commands.
then trying to list boots.
With old systemd v231, same journal files can be parsed correctly, and thus boot ids are listed corrected. But with systemd v232,v233,v234, there comes error, and no boot ids are listed.
binary journal files cannot be shared because of company confidentially. Maybe if this can be produced in plain fedora build, i can then share the files.

### PR DESCRIPTION
**Pull request by rantala**
Introduce journal_file_check_object(), which does lightweight object sanity checks, and use it in journal_file_move_to_object(), so that we will catch certain corrupted objects in the journal file.

This fixes #6447, where we had only partially written out OBJECT_ENTRY (ObjectHeader written, but rest of object zero bytes), causing "journalctl --list-boots" to fail.

```
  $ builddir.vanilla/journalctl --list-boots -D bug6447/
  Failed to determine boots: No data available

  $ builddir.patched/journalctl --list-boots -D bug6447/
  -52 22633da1c5374a728d6c215e2c301dc2 Mon 2017-07-10 05:29:21 EEST—Mon 2017-07-10 05:31:51 EEST
  -51 2253aab9ea7e4a2598f2abda82939eff Mon 2017-07-10 05:32:22 EEST—Mon 2017-07-10 05:36:49 EEST
  -50 ef0d85d35c74486fa4104f9d6391b6ba Mon 2017-07-10 05:40:33 EEST—Mon 2017-07-10 05:40:40 EEST
  [...]
```

Note that journal_file_check_object() is similar to journal_file_object_verify(). The most expensive checks are omitted, as they would slow down every journal_file_move_to_object() call too much.

With this implementation, the added overhead is small, for example when dumping some journal content to /dev/null (built with -Dbuildtype=debugoptimized -Db_ndebug=true):

```
 Performance counter stats for 'builddir.vanilla/journalctl -D 76f4d4c3406945f9a60d3ca8763aa754/':

      12542,311634      task-clock:u (msec)       #    1,000 CPUs utilized
                 0      context-switches:u        #    0,000 K/sec
                 0      cpu-migrations:u          #    0,000 K/sec
            80 100      page-faults:u             #    0,006 M/sec
    41 786 963 456      cycles:u                  #    3,332 GHz
   105 453 864 770      instructions:u            #    2,52  insn per cycle
    24 342 227 334      branches:u                # 1940,809 M/sec
       105 709 217      branch-misses:u           #    0,43% of all branches

      12,545199291 seconds time elapsed

 Performance counter stats for 'builddir.patched/journalctl -D 76f4d4c3406945f9a60d3ca8763aa754/':

      12734,723233      task-clock:u (msec)       #    1,000 CPUs utilized
                 0      context-switches:u        #    0,000 K/sec
                 0      cpu-migrations:u          #    0,000 K/sec
            80 693      page-faults:u             #    0,006 M/sec
    42 661 017 429      cycles:u                  #    3,350 GHz
   107 696 985 865      instructions:u            #    2,52  insn per cycle
    24 950 526 745      branches:u                # 1959,252 M/sec
       101 762 806      branch-misses:u           #    0,41% of all branches

      12,737527327 seconds time elapsed
```

Fixes #6447.

### CODE REVIEW

**Comment by poettering** [comment]
Oh, wow, this is excellent. Thanks for doing this, even for the different object types!

File: src/journal/journal-file.c [hunk 1]
```diff
@@ -770,6 +770,169 @@ static uint64_t minimum_header_size(Object *o) {
...
+                        log_debug(
+                              "Bad field size (<= %zu): %"PRIu64": %"PRIu64,
+                              offsetof(FieldObject, payload),
+                              le64toh(o->object.size),
+                              offset);
+                        return -EBADMSG;
+                }
+
+                if (!VALID64(o->field.next_hash_offset) ||
+                    !VALID64(o->field.head_data_offset)) {
```
Comment (poettering) [inline src/journal/journal-file.c]: hmm, this is missing le64toh(). everything in the journal files is LE after all. And this breaks s390 apparently...

Hmm, so, apparently journal-verify.c already has a couple of missing le64toh...

**Comment by poettering** [comment]
So I have prepped a patch in #6888, that fixes the LE/BE issues elsewhere, can you make similar fixes on your patch?

**Comment by rantala** [comment]
> So I have prepped a patch in #6888, that fixes the LE/BE issues elsewhere, can you make similar fixes on your patch?

Cool, I'll update this patch!

**Comment by rantala** [comment]
OK, added the missing le64toh() calls in a second commit.

**Comment by poettering** [comment]
CI failure appears unrelated. Thanks! Merging!

**Pull request merged**